### ISSUE TITLE
Three mousewheel scrolls needed to zoom

### ISSUE THREAD

**Issue by legendre6891** [moosotc/llpp#59]
When holding `ctrl`, I found that it requires three 'ticks' of scrolling on the mousewheel to trigger a zoom-in or zoom-out. Is this intentional? If so, is this configurable so that only one tick does the job?

**Comment by moosotc** [comment]
Kevin Li <notifications@github.com> writes:

> When holding `ctrl`, I found that it requires three 'ticks' of
> scrolling on the mousewheel to trigger a zoom-in or zoom-out. Is this
> intentional? If so, is this configurable so that only one tick does
> the job?

It was intentional (i.e. the code went to some lengths to achieve that),
however I can't recall why. Following should "fix" that. (Replacing
`if i = 2' with `if true' in main.ml:5707 should have the same effect)

diff --git a/config.ml b/config.ml
index 407b937..957bc81 100644
--- a/config.ml
+++ b/config.ml
@@ -336,7 +336,6 @@ type mpos = int * int
    | Msel of (mpos * mpos)
    | Mpan of mpos
    | Mscrolly | Mscrollx
-   | Mzoom of (buttonno * step * mpos)
    | Mzoomrect of (mpos * mpos)
    | Mnone
  and buttonno = int
diff --git a/main.ml b/main.ml
index 4e9ecd4..d4b0087 100644
--- a/main.ml
+++ b/main.ml
@@ -3003,7 +3003,6 @@ object (self)
     | Msel _
     | Mpan _
     | Mscrollx
-    | Mzoom _
     | Mzoomrect _
     | Mnone -> coe self
 
@@ -4718,7 +4717,6 @@ let viewkeyboard key mask =
      | Msel _
      | Mpan _
      | Mscrolly | Mscrollx
-     | Mzoom _
      | Mnone ->
         begin match state.mode with
         | LinkNav ln ->
@@ -5481,7 +5479,7 @@ let scrollindicator () =
 
 let showsel () =
   match state.mstate with
-  | Mnone | Mscrolly | Mscrollx | Mpan _ | Mzoom _ | Mzoomrect _ ->
+  | Mnone | Mscrolly | Mscrollx | Mpan _ | Mzoomrect _ ->
      ()
 
   | Msel ((x0, y0), (x1, y1)) ->
@@ -5561,7 +5559,6 @@ let display () =
   | Msel _
   | Mpan _
   | Mscrolly | Mscrollx
-  | Mzoom _
   | Mnone -> ()
   end;
   enttext ();
@@ -5695,39 +5692,13 @@ let viewmouse button down x y mask =
   | n when (n == 4 || n == 5) && not down ->
      if Wsi.withctrl mask
      then (
-       match state.mstate with
-       | Mzoom (oldn, i, (ftx, fty)) ->
-          let recenter =
-            if false
-            then abs (ftx - x) > 5 || abs (fty - y) > 5
-            else false
-          in
-          if oldn = n
-          then (
-            if i = 2
-            then
-              let incr =
-                match n with
-                | 5 ->
-                   if conf.zoom +. 0.01 > 0.1 then 0.1 else 0.01
-                | _ ->
-                   if conf.zoom -. 0.1 < 0.1 then -0.01 else -0.1
-              in
-              let zoom = conf.zoom -. incr in
-              if recenter
-              then pivotzoom ~x ~y zoom
-              else pivotzoom zoom;
-              state.mstate <- Mzoom (n, 0, (x, y));
-            else
-              state.mstate <- Mzoom (n, i+1, (ftx, fty));
-          )
-          else state.mstate <- Mzoom (n, 0, (ftx, fty))
-
-       | Msel _
-       | Mpan _
-       | Mscrolly | Mscrollx
-       | Mzoomrect _
-       | Mnone -> state.mstate <- Mzoom (n, 0, (0, 0))
+       let incr =
+         match n with
+         | 5 -> if conf.zoom +. 0.01 > 0.1 then 0.1 else 0.01
+         | _ -> if conf.zoom -. 0.1 < 0.1 then -0.01 else -0.1
+       in
+       let zoom = conf.zoom -. incr in
+       pivotzoom zoom;
      )
      else (
        match state.autoscroll with
@@ -5807,7 +5778,6 @@ let viewmouse button down x y mask =
        | Msel _
        | Mpan _
        | Mscrolly | Mscrollx
-       | Mzoom _
        | Mnone -> resetmstate ()
      )
 
@@ -5858,7 +5828,7 @@ let viewmouse button down x y mask =
           match state.mstate with
           | Mnone -> ()
 
-          | Mzoom _ | Mscrollx | Mscrolly ->
+          | Mscrollx | Mscrolly ->
              state.mstate <- Mnone
 
           | Mzoomrect ((x0, y0), _) ->
@@ -5954,7 +5924,7 @@ let uioh = object
       | Textentry _ -> ()
       | View | Birdseye _ | LinkNav _ ->
          match state.mstate with
-         | Mzoom _ | Mnone -> ()
+         | Mnone -> ()
 
          | Mpan (x0, y0) ->
             let dx = x - x0
@@ -6007,7 +5977,7 @@ let uioh = object
 
       | LinkNav _ | View ->
          match state.mstate with
-         | Mpan _ | Msel _ | Mzoom _ | Mscrolly | Mscrollx | Mzoomrect _ -> ()
+         | Mpan _ | Msel _ | Mscrolly | Mscrollx | Mzoomrect _ -> ()
          | Mnone ->
             updateunder x y;
             if canselect ()


-- 
mailto:moosotc@gmail.com

**Comment by legendre6891** [comment]
Ah thanks! Perhaps the original reason for this change is that 3 'scrolls' on a trackpad may be reasonable (compared to three scrolls on a mousewheel).

Note that changing `i=2` to `true` produces this funny behavior:

1. First zoom-out using `ctrl` + mousewheel up
2. Now attempt to zoom-in by using `ctrl` and mousewheel down. *This first mousewheel down event now no effect.* 
3. After the first mousewheel down, zoom-in happens once per scroll, as expected.

Basically, when changing directions of zoom, the first mousewheel tick is lost. 


Since you showed me the diff I'll see if I can track down how this is happening (unless you already know why). I'll report back.

**Comment by moosotc** [comment]
Kevin Li <notifications@github.com> writes:

> Ah thanks! Perhaps the original reason for this change is that 3
> 'scrolls' on a trackpad may be reasonable (compared to three scrolls
> on a mousewheel).
>

I never owned a machine with a trackpad, so that's not it. I was
thinking that the reason was that the wheel on the mouse I had at the
time emitted way too many "clicks", but that's just a conjecture.

> Note that changing `i=2` to `true` produces this funny behavior:
>
> 1. First zoom-out using `ctrl` + mousewheel up
> 2. Now attempt to zoom-in by using `ctrl` and mousewheel down. *This
> first mousewheel down event now no effect.*
> 3. After the first mousewheel down, zoom-in happens once per scroll,
> as expected.
>
> Basically, when changing directions of zoom, the first mousewheel tick is lost. 

It's lost when initiating the zoom too, that's just how the code is
(mis)written.

>
>
> Since you showed me the diff I'll see if I can track down how this is
> happening (unless you already know why). I'll report back.

Bellow is the thing that shouldn't be susceptible to this. I'm inclined
to commit it.

diff --git a/main.ml b/main.ml
index 4e9ecd4..7ab6d43 100644
--- a/main.ml
+++ b/main.ml
@@ -5695,39 +5695,22 @@ let viewmouse button down x y mask =
   | n when (n == 4 || n == 5) && not down ->
      if Wsi.withctrl mask
      then (
-       match state.mstate with
-       | Mzoom (oldn, i, (ftx, fty)) ->
-          let recenter =
-            if false
-            then abs (ftx - x) > 5 || abs (fty - y) > 5
-            else false
-          in
-          if oldn = n
-          then (
-            if i = 2
-            then
-              let incr =
-                match n with
-                | 5 ->
-                   if conf.zoom +. 0.01 > 0.1 then 0.1 else 0.01
-                | _ ->
-                   if conf.zoom -. 0.1 < 0.1 then -0.01 else -0.1
-              in
-              let zoom = conf.zoom -. incr in
-              if recenter
-              then pivotzoom ~x ~y zoom
-              else pivotzoom zoom;
-              state.mstate <- Mzoom (n, 0, (x, y));
-            else
-              state.mstate <- Mzoom (n, i+1, (ftx, fty));
-          )
-          else state.mstate <- Mzoom (n, 0, (ftx, fty))
-
-       | Msel _
-       | Mpan _
-       | Mscrolly | Mscrollx
-       | Mzoomrect _
-       | Mnone -> state.mstate <- Mzoom (n, 0, (0, 0))
+       let incr =
+         if n = 4
+         then if conf.zoom +. 0.01 > 0.1 then 0.1 else 0.01
+         else if conf.zoom -. 0.1 < 0.1 then -0.01 else -0.1
+       in
+       let fx, fy =
+         match state.mstate with
+         | Mzoom (oldn, _, pos) when n = oldn -> pos
+         | Mzoomrect _ | Mnone | Mpan _
+         | Msel _ | Mscrollx | Mscrolly | Mzoom _ -> (x, y)
+       in
+       let zoom = conf.zoom -. incr in
+       state.mstate <- Mzoom (n, 0, (x, y));
+       if false && abs (fx - x) > 5 || abs (fy - y) > 5
+       then pivotzoom ~x ~y zoom
+       else pivotzoom zoom
      )
      else (
        match state.autoscroll with

-- 
mailto:moosotc@gmail.com
